From 8c1e8f30beac6008e6359fa61593e6f8ba1c5622 Mon Sep 17 00:00:00 2001
From: Nello De Gregoris <bluechxindv@gmail.com>
Date: Tue, 22 Sep 2026 17:58:10 +0200
Subject: [PATCH] rsaenh: Return NTE_BAD_ALGID for an HMAC with no hash
 algorithm.

(cherry picked from commit 4d67b19f26ea217096609a03ae7cf5219d5db1df)

CW-Bug-Id: #27698
---
 dlls/rsaenh/rsaenh.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/dlls/rsaenh/rsaenh.c b/dlls/rsaenh/rsaenh.c
index b0555db4b90..b696fed529d 100644
--- a/dlls/rsaenh/rsaenh.c
+++ b/dlls/rsaenh/rsaenh.c
@@ -3764,6 +3764,12 @@ BOOL WINAPI RSAENH_CPGetHashParam(HCRYPTPROV hProv, HCRYPTHASH hHash, DWORD dwPa
                                 &pCryptHash->tpPRFParams.blobSeed, pbData, *pdwDataLen);
             }
 
+            if (pCryptHash->aiAlgid == CALG_HMAC && !pCryptHash->hash.desc)
+            {
+                SetLastError(NTE_BAD_ALGID);
+                return FALSE;
+            }
+
             if (pCryptHash->dwState != RSAENH_HASHSTATE_FINISHED)
             {
                 finalize_hash(pCryptHash);
@@ -4621,6 +4627,12 @@ BOOL WINAPI RSAENH_CPHashData(HCRYPTPROV hProv, HCRYPTHASH hHash, const BYTE *pb
         SetLastError(NTE_BAD_ALGID);
         return FALSE;
     }
+
+    if (pCryptHash->aiAlgid == CALG_HMAC && !pCryptHash->hash.desc)
+    {
+        SetLastError(NTE_BAD_ALGID);
+        return FALSE;
+    }
     
     if (pCryptHash->dwState != RSAENH_HASHSTATE_HASHING)
     {
